### PR TITLE
[SPARK-34736][K8S][TESTS] Kubernetes and Minikube version upgrade for integration tests

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

This PR upgrades Kubernetes and Minikube version for integration tests and removes/updates the old code for this new version.

Details of this changes: 

- As [discussed in the mailing list](http://apache-spark-developers-list.1001551.n3.nabble.com/minikube-and-kubernetes-cluster-versions-for-integration-testing-td30856.html): updating Minikube version from v0.34.1 to v1.7.3 and kubernetes version from v1.15.12 to v1.17.3.
- making Minikube version checked and fail with an explanation when the test is started with on a version <  v1.7.3.
- removing minikube status checking code related to old Minikube versions 
- in the Minikube backend using fabric8's `Config.autoConfigure()` method to configure the kubernetes client to use the `minikube` k8s context (like it was in [one of the Minikube's example](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/kubectl/equivalents/ConfigUseContext.java#L36))
- Introducing `persistentVolume` test tag: this would be a temporary change to skip PVC tests in the Kubernetes integration test, as currently the PCV tests are blocking the move to Docker as Minikube's driver (for details please check https://issues.apache.org/jira/browse/SPARK-34738). 

### Why are the changes needed?

With the current suggestion one can run into several problems without noticing the Minikube/kubernetes version is the problem.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

It was tested on Mac with [this script](https://gist.github.com/attilapiros/cd58a16bdde833c80c5803c337fffa94#file-check_minikube_versions-zsh) which installs each Minikube versions from v1.7.2 (including this version to test the negative case of the version check) and runs the integration tests.

It was started with:
```
./check_minikube_versions.zsh > test_log 2>&1
```

And there was only one build failure the rest was successful:

```
$ grep "BUILD SUCCESS" test_log | wc -l
      26
$ grep "BUILD FAILURE" test_log | wc -l
       1
```

It was for Minikube v1.7.2  and the log is:

```
KubernetesSuite:
*** RUN ABORTED ***
  java.lang.AssertionError: assertion failed: Unsupported Minikube version is detected: minikube version: v1.7.2.For integration testing Minikube version 1.7.3 or greater is expected.
  at scala.Predef$.assert(Predef.scala:223)
  at org.apache.spark.deploy.k8s.integrationtest.backend.minikube.Minikube$.getKubernetesClient(Minikube.scala:52)
  at org.apache.spark.deploy.k8s.integrationtest.backend.minikube.MinikubeTestBackend$.initialize(MinikubeTestBackend.scala:33)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.beforeAll(KubernetesSuite.scala:163)
  at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:212)
  at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
  at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.org$scalatest$BeforeAndAfter$$super$run(KubernetesSuite.scala:43)
  at org.scalatest.BeforeAndAfter.run(BeforeAndAfter.scala:273)
  at org.scalatest.BeforeAndAfter.run$(BeforeAndAfter.scala:271)
  ...
```

Moreover I made a test with having multiple k8s cluster contexts, too.

### CODE REVIEW

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala
Comment (rvesse) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:61]: If we're now enforcing that the name of the context must be `minikube` this should be included in the apache/spark-website#321 PR

**Review by rvesse — APPROVED** [review]
LGTM, one comment about needing a doc update around the requirement for the context to be called `minikube`

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala
Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:61]: I think that is the default context for minikube. Although I have not found this in documentation so I asked this in the Minikube user's list:

```
Hi!

As I see the default context for the started cluster is `minikube` but I have not found this in documentation.
- Am I right? 
- Can I really on this?
- Could you please show me the documentation about this (or the relevant code part)?

Thanks in advance,
Attila
```

I hope we can get an answer soon.

Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:61]: I'm sure it's probably is the default (though their community will be best placed to confirm).  The scenario I was thinking of was more the one where users have renamed their contexts to their own preference.  I have access to a lot of R&D clusters in my $dayjob and at times I've had as many as 10+ different contexts in my `KUBECONFIG` so I would rename contexts appropriately as some tools (GKE I'm looking at you) produce rather user unfriendly names by default.  Not sure I've ever renamed the `minikube` context itself but it's a thing that users can and might do.

TL;DR If there's a naming assumption present in the code, particularly where it pertains to user managed configuration files outside of Spark's control, it should be documented as such.

Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:61]: I have tested this case and when after minikube is started and the context is switched to another one even that makes the test fail. 

So the best probably is mentioning that fact too:
```
- the current kubernetes context must be minikube's default context (called 'minikube'). This can be selected by `minikube kubectl -- config use-context minikube`. This is only needed when after minikube is started another kubernetes context is selected.
```

**Comment by fbalicchia** [comment]
Hi, I'm experiencing in the same issue and and to resolve the issue I've applied the same changes. Could you please merge it ?

**Comment by attilapiros** [comment]
@fbalicchia Could you please share with us the error you run into (the one on which this PR helped on)?
By the way I cannot merge my own PR without an expert's approval (expert of the area where the PR belongs to). But your input will help the case.

**Comment by fbalicchia** [comment]
Using minikube `v1.18.1` default driver is Docker and at the moment of writing masterUrl not necessary bind on port `8443` https://github.com/apache/spark/blob/master/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala#L69.
Instead using default config or using factory with `autoConfigure("minikube")` library help us

File: resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh
Comment (attilapiros) [inline resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh:183]: The `install` phase is after `integration-test`, see default lifecycle at the [maven doc](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html):


Phase | Description
-- | --
... | ...
package | take the compiled code and package it in its distributable format, such as a JAR.
pre-integration-test | perform actions required before integration tests are executed. This may involve things such as setting up the required environment.
**integration-test** | process and deploy the package if necessary into an environment where integration tests can be run.
post-integration-test | perform actions required after integration tests have been executed. This may including cleaning up the environment.
verify | run any checks to verify the package is valid and meets quality criteria.
**install** | install the package into the local repository, for use as a dependency in other projects locally.
...|...


With `install` the local repo will be updated with the fresh jars and in case of `--skip-building-dependencies` the build will reuse those.

**Comment by attilapiros** [comment]
rebased on master

**Review by dongjoon-hyun — COMMENTED** [review]
@attilapiros . Is there an ETA for the blocker you mentioned? If there is no ETA for that, it would be great if we can proceed a subset first partially. For those parts which are out of our control, we had better reduce the bottleneck or dependency as much as possible.

BTW, Spark 3.2 code freeze is `July 1st 2021`. So, we still have enough dev time though.

**Comment by attilapiros** [comment]
@dongjoon-hyun actually I am pretty sure this helps (even with the skipping of the PVTests which when @shaneknapp sets the Minikube driver to Docker will fix the Kubernetes integration tests).

So could you please review this as it is?

**Comment by dongjoon-hyun** [comment]
Oh, I missed your last comment here. Very sorry, @attilapiros .

File: resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh:183]: Well, I know that we need `install` sometimes, but for this one, I'm a little negative for converting `integration-test` to `install` because it pollutes the local repo while testing and reviewing someone-else's PR. It's really painful for committers to clean up the local repo everytime to avoid any side-effects for the other PRs. Is this really inevitable?

Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh:183]: In addition, this is inconsistent from README.md line 128.
- https://github.com/apache/spark/pull/31829/files#diff-280a1f3f6bebe922254b2a1f2f952929f02e582c60169c659f1dff792f633caaR128

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:33]: Most of changes are only related to these spacing changes. I'm wondering if these changes are inevitable for the purpose of this PR.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:568]: Is this new tag documented or mentioned somewhere in this PR?

**Comment by dongjoon-hyun** [comment]
I left a few comments.
- https://github.com/apache/spark/pull/31829#discussion_r628834835
- https://github.com/apache/spark/pull/31829#discussion_r628835311
- https://github.com/apache/spark/pull/31829#discussion_r628835797
- https://github.com/apache/spark/pull/31829#discussion_r628835907

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:568]: This would be a temporary change to enable Jenkins to use Docker as minikube's driver.
This way Kubernetes integration test would be successful.

More about it:
https://issues.apache.org/jira/browse/SPARK-34738?focusedCommentId=17324165&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17324165

File: resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh
Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh:183]: I would say this is the orherway around, assuming in both PRs we stick to `install`.
So when we call `install` in the first PR we could be sure the fresh jars will be used (even when the artifact we build referencing something outside of the maven reactor) and when we move to the next PR because of `install` we have again the fresh jars.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala
Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:33]: I can revert them but I think this is more clean than polluting each usage with the extra space. In addition we can use the `split` method (and getting the second part) instead of the `replace_first` which express our intention a bit better.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:568]: Yes, that's exactly what I mean. We had better mention this workaround and limitation (ignoring PVC tests) of this PR in the description for the reviewers.
> This would be a temporary change to enable Jenkins to use Docker as minikube's driver.
This way Kubernetes integration test would be successful.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala:33]: Then, if you don't mind, shall you split that refactoring contribution into another PR? Each PR had better have one theme and minimal.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (attilapiros) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:568]: I got it :)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM Thank you, @attilapiros and all.

Sorry for back and forth. On second thought, for the above all my comments, we can handle them later when we need them. Let's merge this and move forward. Thanks!

**Pull request closed without merging**
